tools/ulog_gen.py

```diff
@@ -0,0 +1,284 @@
+
+    Message_Header_Type header = {
+        .msg_type = 'A',
+        .msg_size = sizeof(multi_id) + sizeof(log->next_msg_id) + strlen(name),
+    };
+
+    int ret = fs_write(&log->file, &header, sizeof(header));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, &multi_id, sizeof(multi_id));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, &log->next_msg_id, sizeof(log->next_msg_id));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, name, strlen(name));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    *msg_id = log->next_msg_id;
+    log->next_msg_id++;
+
+    return ULOG_SUCCESS;
+}
+
+ULOG_Error_Type ULOG_{{ struct_name }}_Write(ULOG_Inst_Type *log, const ULOG_{{ struct_name }}_Type* {{ struct_name_lower }}, const uint16_t msg_id) {
+    if (log == NULL || {{ struct_name_lower }} == NULL) {
+        return ULOG_INVALID_PARAM;
+    }
+
+    if (log->phase != ULOG_PHASE_DATA) {
+        return ULOG_WRONG_PHASE;
+    }
+
+    const uint64_t uptime_us = k_uptime_get() * 1000;
+
+    Message_Header_Type header = {
+        .msg_type = 'D',
+        .msg_size = sizeof(msg_id)
+            + sizeof(uptime_us)
+            {%- for field in fields %}
+            {% if field.array_length %}({{+ field.array_length }} * sizeof({{ field.type }})){% else %}+ sizeof({{ field.type }}){% endif %}
+            {%- endfor %},
+    };
+
+    // Write the message metadata
+    int ret = fs_write(&log->file, &header, sizeof(header));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, &msg_id, sizeof(msg_id));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    ret = fs_write(&log->file, &uptime_us, sizeof(uptime_us));
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+
+    // Write the message data
+    {%- for field in fields %}
+    ret = fs_write(&log->file, {% if field.array_length %}{{ struct_name_lower }}->{{ field.name }}{% else %}&{{ struct_name_lower }}->{{ field.name }}{% endif %}, {% if field.array_length %}{{ field.array_length }} * sizeof({{ field.type }}){% else %}sizeof({{ field.type }}){% endif %});
+    if (ret < 0) {
+        return ULOG_FILESYSTEM_ERROR;
+    }
+    {% endfor %}
+    return ULOG_SUCCESS;
+}
+
+"""
+
+def capitalize_snake_case(value):
+    words = value.split('_')
+    return '_'.join(word.capitalize() for word in words)
+
+def check_field_types(message):
+    valid_types = ["int8_t", "uint8_t", "int16_t", "uint16_t", "int32_t", "uint32_t",
+                   "int64_t", "uint64_t", "float", "double", "char", "bool"]
+
+    for field in message["fields"]:
+        if field["type"] not in valid_types:
+            raise SystemExit(f"Field {field["name"]} in {message["name"]} message is of unsupported type - {field["type"]}")
```

Comment:
Error is generated while building the project using west
![Screenshot from 2024-12-16 17-07-04](https://github.com/user-attachments/assets/a446d2f4-d2aa-4f74-bf5f-7c605b7df5d6)

It appears that the issue is because of usage of nested double quotation marks in f-string
Issue can be fixed by replacing outer double quotation marks with '

Reply:
Interesting, I haven't encountered this error at all, but you are right, thanks very much for finding this!